### PR TITLE
feat(Core/World): Implement DoForAllPlayers() helper for all online p…

### PR DESCRIPTION
**Pull request by Nyeriah**
…layers

## Changes Proposed:

This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes 

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [ ] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1.
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by avarishd** [comment]
Maybe DoForAllOnlinePlayers, to not share name with map one?

File: src/server/game/World/World.cpp [hunk 1]
```diff
@@ -3256,6 +3256,25 @@ void World::RemoveOldCorpses()
     _timers[WUPDATE_CORPSES].SetCurrent(_timers[WUPDATE_CORPSES].GetInterval());
 }
 
+void World::DoForAllPlayers(std::function<void(Player*)> exec)
+{
+    std::shared_lock<std::shared_mutex> lock(*HashMapHolder<Player>::GetLock());
```
Comment (Winfidonarleyan) [inline src/server/game/World/World.cpp:3261]:
```suggestion
    std::shared_lock lock(*HashMapHolder<Player>::GetLock());
```

Comment (Winfidonarleyan) [reply src/server/game/World/World.cpp:3261]: https://en.cppreference.com/w/cpp/language/template_argument_deduction

File: src/server/game/World/World.cpp [hunk 2]
```diff
@@ -3256,6 +3256,25 @@ void World::RemoveOldCorpses()
...
+    const HashMapHolder<Player>::MapType& playerMap = ObjectAccessor::GetPlayers();
+
+    for (auto const& it : playerMap)
```
Comment (Winfidonarleyan) [inline src/server/game/World/World.cpp:3264]:
```suggestion
    for (auto const& it : ObjectAccessor::GetPlayers())
```

File: src/server/scripts/Commands/cs_reset.cpp [hunk 3]
```diff
@@ -290,10 +290,9 @@ class reset_commandscript : public CommandScript
         stmt->SetData(0, uint16(atLogin));
         CharacterDatabase.Execute(stmt);
 
-        std::shared_lock<std::shared_mutex> lock(*HashMapHolder<Player>::GetLock());
-        HashMapHolder<Player>::MapType const& plist = ObjectAccessor::GetPlayers();
-        for (auto itr = plist.begin(); itr != plist.end(); ++itr)
-            itr->second->SetAtLoginFlag(atLogin);
+        sWorld->DoForAllPlayers([&] (Player* player){
+            player->SetAtLoginFlag(atLogin);
+        });
```
Comment (Winfidonarleyan) [inline src/server/scripts/Commands/cs_reset.cpp:295]:
```suggestion
        sWorld->DoForAllPlayers([&] (Player* player)
        {
            player->SetAtLoginFlag(atLogin);
        });
```

**Comment by Winfidonarleyan** [comment]
> Maybe DoForAllOnlinePlayers, to not share name with map one?

Good case, maybe make `void World::DoForAllPlayers(std::function<void(Player*)> exec, bool isOnline = false)` ?

**Comment by Nyeriah** [comment]
> > Maybe DoForAllOnlinePlayers, to not share name with map one?
> 
> Good case, maybe make, example:
> 
> ```c++
> void World::DoForAllPlayers(std::function<void(Player*)> exec, bool isOnline /*= false*/)
> ```

We use the session list, so we can only iterate through online players

**Review by elthehablo — APPROVED** [review]

**Pull request merged**